src/NServiceBus.Core/Transports/Learning/LearningTransportDispatcher.cs

```diff
@@ -66,23 +69,23 @@ await AsyncFile.WriteBytes(bodyPath, transportOperation.Message.Body)
                 .ConfigureAwait(false);
 
             DateTime? timeToDeliver = null;
-            DelayDeliveryWith delayDeliveryWith;
 
-            if (transportOperation.DeliveryConstraints.TryGet(out delayDeliveryWith))
+            if (transportOperation.DeliveryConstraints.TryGet(out DoNotDeliverBefore doNotDeliverBefore))
             {
-                timeToDeliver = DateTime.UtcNow + delayDeliveryWith.Delay;
+                timeToDeliver = doNotDeliverBefore.At;
             }
-
-            DoNotDeliverBefore doNotDeliverBefore;
-
-            if (transportOperation.DeliveryConstraints.TryGet(out doNotDeliverBefore))
+            else if (transportOperation.DeliveryConstraints.TryGet(out DelayDeliveryWith delayDeliveryWith))
             {
-                timeToDeliver = doNotDeliverBefore.At;
+                timeToDeliver = DateTime.UtcNow + delayDeliveryWith.Delay;
             }
 
-
             if (timeToDeliver.HasValue)
             {
+                if (transportOperation.DeliveryConstraints.TryGet(out DiscardIfNotReceivedBefore timeToBeReceived) && timeToBeReceived.MaxTime < TimeSpan.MaxValue)
+                {
+                    throw new Exception("Postponed delivery of messages with TimeToBeReceived set is not supported. Remove the TimeToBeReceived attribute to postpone messages of this type.");
```

Comment:
You mean to show the value of the enclosed message types header? (since that all we got at this low level)